packages/apollo-federation/src/composition/validate/sdl/uniqueTypeNamesWithoutEnumsOrScalars.ts

```diff
@@ -30,25 +50,128 @@ export function UniqueTypeNamesWithoutEnumsOrScalars(
   };
 
-  function checkTypeName(node: TypeDefinitionNode) {
+  function checkTypeName(node: TypesWithRequiredUniqueNames) {
     const typeName = node.name.value;
+    const typeFromSchema = schema && schema.getType(typeName);
+    const typeNodeFromSchema =
+      typeFromSchema &&
+      (typeFromSchema.astNode as Maybe<TypesWithRequiredUniqueNames>);
+
+    const typeNodeFromDefs = knownTypes[typeName];
+    const duplicateTypeNode = typeNodeFromSchema || typeNodeFromDefs;
+
+    // Handle union types with the same name
+    if (
+      duplicateTypeNode &&
+      duplicateTypeNode.kind === Kind.UNION_TYPE_DEFINITION &&
+      node.kind === Kind.UNION_TYPE_DEFINITION
+    ) {
+      const unionDiff = xorBy(
+        node.types,
+        duplicateTypeNode.types,
+        'name.value',
+      );
+
+      const diffLength = unionDiff.length;
+      if (diffLength > 0) {
+        context.reportError(
+          errorWithCode(
+            'VALUE_TYPE_UNION_TYPES_MISMATCH',
+            `The union '${typeName}' is defined in multiple places, however the unioned types do not match. Union types with the same name must also consist of identical types. The type${
+              diffLength > 1 ? 's' : ''
+            } ${unionDiff.map(diffEntry => diffEntry.name.value).join(', ')} ${
+              diffLength > 1 ? 'are' : 'is'
+            } mismatched.`,
+            [node, duplicateTypeNode],
+          ),
+        );
+      }
+
+      return false;
+    }
+
+    /*
+     * Return early for value types
+     * Value types:
+     * 1) have the same kind (type, interface, input), extensions are excluded
+     * 2) are not entities
+     * 3) have the same set of fields
+     */
+    if (duplicateTypeNode) {
+      const possibleErrors: GraphQLError[] = [];
+      // By inspecting the diff, we can warn when field types mismatch.
+      // A diff entry will exist when a field exists on one type and not the other, or if there is a type mismatch on the field
+      // i.e. { sku: [Int, String!], color: [String] }
+      const { kind, fields } = diffTypeNodes(node, duplicateTypeNode);
+
+      const fieldsDiff = Object.entries(fields);
+      const typesHaveSameShape =
+        fieldsDiff.length === 0 ||
+        fieldsDiff.every(([fieldName, types]) => {
+          // If a diff entry has two types, then the field name matches but the types do not.
+          // In this case, we can push a useful error to hint to the user that we
+          // think they tried to define a value type, but one of the fields has a type mismatch.
+          if (types.length === 2) {
+            possibleErrors.push(
+              errorWithCode(
+                'VALUE_TYPE_FIELD_TYPE_MISMATCH',
+                `Found field type mismatch on expected value type. '${typeName}.${fieldName}' is defined as both a ${types[0]} and a ${types[1]}. In order to define '${typeName}' in multiple places, the fields and their types must be identical.`,
```

Comment:
The wording of this error is hard for me to grasp. Can we word it to show something like this?


```
[ServiceA] Product.color: A field was defined differently in different services. ServiceA and ServiceB define Product.Color as a String! and String respectively. In order to define Product in multiple places, the fields and their types must be identical.
```

An example of an error like this is here: https://github.com/apollographql/apollo-server/blob/master/packages/apollo-federation/src/composition/validate/sdl/possibleTypeExtensions.ts#L67

Reply:
Your suggestion is great 👍
Resolved in 8fc61bdf58635c81bd6fe1b4930568560ce676f0